### PR TITLE
expansion multiply

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

File: dpnp/backend_mathematical.pyx [hunk 1]
```diff
@@ -159,8 +159,37 @@ cpdef tuple dpnp_modf(dparray x1):
...
+    x2_is_scalar = dpnp.isscalar(x2)
+
+    x1_dtype = x1.dtype
+    x2_dtype = type(x2) if x2_is_scalar else x2.dtype
+
+    if x1_dtype == numpy.float64:
+        res_type = x1_dtype
+    elif x1_dtype == numpy.float32:
+        res_type = x1_dtype
+    if x1_dtype == numpy.int64:
```
Comment (densmirn) [inline dpnp/backend_mathematical.pyx:172]: `elif`?

File: dpnp/dpnp_iface_mathematical.py [hunk 2]
```diff
@@ -684,14 +684,26 @@ def multiply(x1, x2, **kwargs):
...
+    if (not use_origin_backend(x1) and (is_x1_dparray or is_x1_scalar)) and \
+            (not use_origin_backend(x2) and (is_x2_dparray or is_x2_scalar)) and \
+            not (is_x1_scalar and is_x2_scalar) and not kwargs:
+
+        if is_x1_scalar:
+            result = dpnp_multiply(x2, x1)
+        else:
+            if is_x1_dparray and is_x2_dparray:
+                if (x1.size != x2.size):
+                    checker_throw_value_error("multiply", "size", x1.size, x2.size)
```
Comment (densmirn) [inline dpnp/dpnp_iface_mathematical.py:699]: Maybe we will fallback to numpy instead of raising exception?

File: dpnp/backend_mathematical.pyx [hunk 3]
```diff
@@ -159,8 +159,37 @@ cpdef tuple dpnp_modf(dparray x1):
...
+            res_type = x1_dtype
+        else:
+            res_type = x2_dtype
+    elif x1_dtype == numpy.int32:
+        if x2_dtype == numpy.float64 or x2_dtype == float or x2_dtype == numpy.float32:
+            res_type = x2_dtype
+        else:
+            res_type = x1_dtype
+    else:
+        res_type = x1_dtype
```
Comment (densmirn) [inline dpnp/backend_mathematical.pyx:183]: You could use the way to define return type from backend something like that:
```python
types_map = defaultdict(dict)  # from collections import defaultdict
types_map[numpy.int32][numpy.int32] = numpy.int32
types_map[numpy.int32][numpy.int64] = numpy.int64
types_map[numpy.int32][numpy.float32] = numpy.float64
types_map[numpy.int32][numpy.float64] = numpy.float64
types_map[numpy.int64][numpy.int32] = numpy.int64
...
res_type = types_map[x1_dtype][x2_dtype]
```

File: dpnp/dpnp_iface_mathematical.py [hunk 4]
```diff
@@ -684,14 +684,19 @@ def multiply(x1, x2, **kwargs):
     is_x1_dparray = isinstance(x1, dparray)
     is_x2_dparray = isinstance(x2, dparray)
 
-    if (not use_origin_backend(x1) and is_x1_dparray and is_x2_dparray and not kwargs):
-        if (x1.size != x2.size):
-            checker_throw_value_error("multiply", "size", x1.size, x2.size)
+    is_x1_scalar = numpy.isscalar(x1)
```
Comment (shssf) [inline dpnp/dpnp_iface_mathematical.py:687]:
```suggestion
    is_x1_scalar = dpnp.isscalar(x1)
```

File: dpnp/backend_mathematical.pyx [hunk 5]
```diff
@@ -159,8 +159,37 @@ cpdef tuple dpnp_modf(dparray x1):
...
 
-cpdef dparray dpnp_multiply(dparray x1, dparray x2):
-    return call_fptr_2in_1out(DPNP_FN_MULTIPLY, x1, x2, x1.shape)
+cpdef dparray dpnp_multiply(dparray x1, x2):
+    x2_is_scalar = dpnp.isscalar(x2)
+
+    x1_dtype = x1.dtype
+    x2_dtype = type(x2) if x2_is_scalar else x2.dtype
+
+    if x1_dtype == numpy.float64:
```
Comment (shssf) [inline dpnp/backend_mathematical.pyx:168]:
```suggestion
    if x1_dtype == dpnp.float64:
```

File: dpnp/backend_mathematical.pyx [hunk 6]
```diff
@@ -162,25 +162,59 @@ cpdef tuple dpnp_modf(dparray x1):
...
+        x1_dtype = dpnp.int64
+    else:
+        x1_dtype = x1_dtype_
+
+    if x2_dtype_ == float:
+        x2_dtype = dpnp.float64
+    elif x2_dtype_ == int:
+        x2_dtype = dpnp.int64
+    else:
+        x2_dtype = x2_dtype_
```
Comment (densmirn) [inline dpnp/backend_mathematical.pyx:180]: You could use dict to simplify the code:
```python
types_map = {float: dpnp.float64, int: dpnp.int64}
x1_dtype = types_map.get(x1_dtype_, x1_dtype_)
x2_dtype = types_map.get(x2_dtype_, x2_dtype_)
```

**Pull request merged**